File: src/main/java/com/hepdd/gtmthings/common/block/machine/multiblock/part/appeng/MEOutputPartMachine.java
package com.hepdd.gtmthings.common.block.machine.multiblock.part.appeng;

import com.gregtechceu.gtceu.api.GTValues;
import com.gregtechceu.gtceu.api.capability.recipe.IO;
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity;
import com.gregtechceu.gtceu.api.machine.MetaMachine;
import com.gregtechceu.gtceu.api.machine.feature.IInteractedMachine;
import com.gregtechceu.gtceu.api.machine.trait.NotifiableFluidTank;
import com.gregtechceu.gtceu.api.machine.trait.NotifiableItemStackHandler;
import com.gregtechceu.gtceu.api.transfer.fluid.CustomFluidTank;
import com.gregtechceu.gtceu.api.transfer.item.CustomItemStackHandler;
import com.gregtechceu.gtceu.common.machine.multiblock.part.DualHatchPartMachine;
import com.gregtechceu.gtceu.integration.ae2.gui.widget.list.AEListGridWidget;
import com.gregtechceu.gtceu.integration.ae2.machine.feature.IGridConnectedMachine;
import com.gregtechceu.gtceu.integration.ae2.machine.trait.GridNodeHolder;
import com.gregtechceu.gtceu.integration.ae2.utils.KeyStorage;
import com.gregtechceu.gtceu.utils.GTMath;

import com.lowdragmc.lowdraglib.gui.widget.LabelWidget;
import com.lowdragmc.lowdraglib.gui.widget.Widget;
import com.lowdragmc.lowdraglib.gui.widget.WidgetGroup;
import com.lowdragmc.lowdraglib.syncdata.annotation.DescSynced;
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted;
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder;
import com.lowdragmc.lowdraglib.utils.Position;

import net.minecraft.MethodsReturnNonnullByDefault;
import net.minecraft.core.Direction;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.fluids.FluidStack;

import appeng.api.config.Actionable;
import appeng.api.networking.IGridNodeListener;
import appeng.api.networking.IManagedGridNode;
import appeng.api.networking.security.IActionSource;
import appeng.api.stacks.AEFluidKey;
import appeng.api.stacks.AEItemKey;
import appeng.api.stacks.AEKey;
import it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.jetbrains.annotations.NotNull;

import java.util.EnumSet;
import java.util.List;

import javax.annotation.ParametersAreNonnullByDefault;

@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
public class MEOutputPartMachine extends DualHatchPartMachine implements IInteractedMachine, IGridConnectedMachine {

    protected static final ManagedFieldHolder MANAGED_FIELD_HOLDER = new ManagedFieldHolder(MEOutputPartMachine.class, DualHatchPartMachine.MANAGED_FIELD_HOLDER);

    @Override
    public ManagedFieldHolder getFieldHolder() {
        return MANAGED_FIELD_HOLDER;
    }

    @DescSynced
    @Getter
    @Setter
    protected boolean isOnline;
    @Persisted
    protected final GridNodeHolder nodeHolder;
    protected final IActionSource actionSource;
    @Getter
    protected Object2LongOpenHashMap<AEKey> returnBuffer = new Object2LongOpenHashMap<>();
    @Persisted
    private KeyStorage internalBuffer;
    @Persisted
    private KeyStorage internalTankBuffer;

    public MEOutputPartMachine(IMachineBlockEntity holder) {
        super(holder, GTValues.LuV, IO.OUT);
        this.nodeHolder = createNodeHolder();
        this.actionSource = IActionSource.ofMachine(nodeHolder.getMainNode()::getNode);
    }

    protected GridNodeHolder createNodeHolder() {
        return new GridNodeHolder(this);
    }

    @Override
    public IManagedGridNode getMainNode() {
        return nodeHolder.getMainNode();
    }

    @Override
    public void onMainNodeStateChanged(IGridNodeListener.State reason) {
        IGridConnectedMachine.super.onMainNodeStateChanged(reason);
        this.updateInventorySubscription();
    }

    @Override
    protected void updateInventorySubscription() {
        if (shouldSubscribe()) {
            autoIOSubs = subscribeServerTick(autoIOSubs, this::autoIO);
        } else if (autoIOSubs != null) {
            autoIOSubs.unsubscribe();
            autoIOSubs = null;
        }
    }

    /////////////////////////////////
    // ***** Machine LifeCycle ****//
    /////////////////////////////////

    @Override
    protected NotifiableItemStackHandler createInventory(Object... args) {
        this.internalBuffer = new KeyStorage();
        return new InaccessibleInfiniteHandler(this);
    }

    @Override
    protected NotifiableFluidTank createTank(int initialCapacity, int slots, Object... args) {
        this.internalTankBuffer = new KeyStorage();
        return new InaccessibleInfiniteTank(this);
    }

    @Override
    public void onLoad() {
        super.onLoad();
        if (isRemote()) return;
    }

    @Override
    public void onMachineRemoved() {
        var grid = getMainNode().getGrid();
        if (grid != null) {
            if (!internalBuffer.isEmpty()) {
                for (var entry : internalBuffer) {
                    grid.getStorageService().getInventory().insert(entry.getKey(), entry.getLongValue(),
                            Actionable.MODULATE, actionSource);
                }
            }
            if (!internalTankBuffer.isEmpty()) {
                for (var entry : internalTankBuffer) {
                    grid.getStorageService().getInventory().insert(entry.getKey(), entry.getLongValue(),
                            Actionable.MODULATE, actionSource);
                }
            }
        }
    }

    /////////////////////////////////
    // ********** Sync ME *********//
    /////////////////////////////////

    protected boolean shouldSubscribe() {
        return isWorkingEnabled() && isOnline() && (!internalBuffer.storage.isEmpty() || !internalTankBuffer.storage.isEmpty());
    }

    @Override
    protected void autoIO() {
        if (!this.shouldSyncME()) return;
        if (this.updateMEStatus()) {
            var grid = getMainNode().getGrid();
            if (grid != null) {
                if (!internalBuffer.isEmpty()) {
                    internalBuffer.insertInventory(grid.getStorageService().getInventory(), actionSource);
                }
                if (!internalTankBuffer.isEmpty()) {
                    internalTankBuffer.insertInventory(grid.getStorageService().getInventory(), actionSource);
                }
            }
            this.updateInventorySubscription();
        }
    }

    // Item Part
    private class InaccessibleInfiniteHandler extends NotifiableItemStackHandler {

        public InaccessibleInfiniteHandler(MetaMachine holder) {
            super(holder, 1, IO.OUT, IO.NONE, ItemStackTransferDelegate::new);
            internalBuffer.setOnContentsChanged(this::onContentsChanged);
        }
    }

    @NoArgsConstructor
    private class ItemStackTransferDelegate extends CustomItemStackHandler {

        // Necessary for InaccessibleInfiniteHandler
        public ItemStackTransferDelegate(Integer integer) {
            super();
        }

        @Override
        public int getSlots() {
            return Short.MAX_VALUE;
        }

        @Override
        public int getSlotLimit(int slot) {
            return Integer.MAX_VALUE;
        }

        @Override
        public ItemStack getStackInSlot(int slot) {
            return ItemStack.EMPTY;
        }

        @Override
        public void setStackInSlot(int slot, ItemStack stack) {
            // NO-OP
        }

        @Override
        public ItemStack insertItem(int slot, ItemStack stack, boolean simulate) {
            var key = AEItemKey.of(stack);
            int count = stack.getCount();
            long oldValue = internalBuffer.storage.getOrDefault(key, 0);
            long changeValue = Math.min(Long.MAX_VALUE - oldValue, count);
            if (changeValue > 0) {
                if (!simulate) {
                    internalBuffer.storage.put(key, oldValue + changeValue);
                    internalBuffer.onChanged();
                }
                return stack.copyWithCount((int) (count - changeValue));
            } else {
                return ItemStack.EMPTY;
            }
        }

        @Override
        public ItemStack extractItem(int slot, int amount, boolean simulate) {
            return ItemStack.EMPTY;
        }
    }

    // Fluid Part
    private class InaccessibleInfiniteTank extends NotifiableFluidTank {

        CustomFluidTank storage;

        public InaccessibleInfiniteTank(MetaMachine holder) {
            super(holder, List.of(new FluidStorageDelegate()), IO.OUT, IO.NONE);
            internalTankBuffer.setOnContentsChanged(this::onContentsChanged);
            storage = getStorages()[0];
        }

        @Override
        public int getTanks() {
            return 128;
        }

        @Override
        public FluidStack getFluidInTank(int tank) {
            return storage.getFluid();
        }

        @Override
        public int getTankCapacity(int tank) {
            return storage.getCapacity();
        }

        @Override
        public boolean isFluidValid(int tank, @NotNull FluidStack stack) {
            return storage.isFluidValid(stack);
        }
    }

    private class FluidStorageDelegate extends CustomFluidTank {

        public FluidStorageDelegate() {
            super(0);
        }

        @Override
        public int getCapacity() {
            return Integer.MAX_VALUE;
        }

        @Override
        public void setFluid(FluidStack fluid) {
            // NO-OP
        }

        @Override
        public int fill(FluidStack resource, FluidAction action) {
            var key = AEFluidKey.of(resource.getFluid(), resource.getTag());
            int amount = resource.getAmount();
            int oldValue = GTMath.saturatedCast(internalBuffer.storage.getOrDefault(key, 0));
            int changeValue = Math.min(Integer.MAX_VALUE - oldValue, amount);
            if (changeValue > 0 && action.execute()) {
                internalBuffer.storage.put(key, oldValue + changeValue);
                internalBuffer.onChanged();
            }
            return changeValue;
        }

        @Override
        public boolean supportsFill(int tank) {
            return false;
        }

        @Override
        public boolean supportsDrain(int tank) {
            return false;
        }

        @Override
        public CustomFluidTank copy() {
            // because recipe testing uses copy transfer instead of simulated operations
            return new FluidStorageDelegate() {

                @Override
                public int fill(FluidStack resource, FluidAction action) {
                    return super.fill(resource, action);
                }
            };
        }
    }

    @Override
    public void onRotated(Direction oldFacing, Direction newFacing) {
        super.onRotated(oldFacing, newFacing);
        getMainNode().setExposedOnSides(EnumSet.of(newFacing));
    }

    @Override
    public boolean isWorkingEnabled() {
        return true;
    }

    @Override
    public Widget createUIWidget() {
        WidgetGroup group = new WidgetGroup(new Position(0, 0));
        // ME Network status
        group.addWidget(new LabelWidget(0, 0, () -> this.isOnline ?
                "gtceu.gui.me_network.online" :
                "gtceu.gui.me_network.offline"));

        group.addWidget(new AEListGridWidget.Item(5, 20, 3, this.internalBuffer));
        group.addWidget(new AEListGridWidget.Fluid(5, 80, 3, this.internalTankBuffer));
        return group;
    }
}

File: src/main/java/com/hepdd/gtmthings/common/block/machine/multiblock/part/appeng/MEOutputPartMachine.kt
package com.hepdd.gtmthings.common.block.machine.multiblock.part.appeng

import net.minecraft.core.Direction

import appeng.api.config.Actionable
import appeng.api.networking.IGridNodeListener
import appeng.api.networking.IManagedGridNode
import appeng.api.networking.security.IActionSource
import appeng.api.stacks.AEKey
import com.gregtechceu.gtceu.api.GTValues
import com.gregtechceu.gtceu.api.capability.recipe.IO
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity
import com.gregtechceu.gtceu.api.machine.feature.IInteractedMachine
import com.gregtechceu.gtceu.api.machine.trait.NotifiableFluidTank
import com.gregtechceu.gtceu.api.machine.trait.NotifiableItemStackHandler
import com.gregtechceu.gtceu.common.machine.multiblock.part.DualHatchPartMachine
import com.gregtechceu.gtceu.integration.ae2.gui.widget.list.AEListGridWidget
import com.gregtechceu.gtceu.integration.ae2.machine.feature.IGridConnectedMachine
import com.gregtechceu.gtceu.integration.ae2.machine.trait.GridNodeHolder
import com.gregtechceu.gtceu.integration.ae2.utils.KeyStorage
import com.hepdd.gtmthings.api.machine.trait.InaccessibleInfiniteHandler
import com.hepdd.gtmthings.api.machine.trait.InaccessibleInfiniteTank
import com.lowdragmc.lowdraglib.gui.widget.LabelWidget
import com.lowdragmc.lowdraglib.gui.widget.Widget
import com.lowdragmc.lowdraglib.gui.widget.WidgetGroup
import com.lowdragmc.lowdraglib.syncdata.annotation.DescSynced
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder
import com.lowdragmc.lowdraglib.utils.Position
import it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap

import java.util.*

open class MEOutputPartMachine(holder: IMachineBlockEntity) :
    DualHatchPartMachine(holder, GTValues.LuV, IO.OUT),
    IInteractedMachine,
    IGridConnectedMachine {

    companion object {
        @JvmStatic
        val MANAGED_FIELD_HOLDER: ManagedFieldHolder =
            ManagedFieldHolder(MEOutputPartMachine::class.java, DualHatchPartMachine.MANAGED_FIELD_HOLDER)
    }

    @DescSynced
    private var isMachineOnline: Boolean = false

    @Persisted
    private var nodeHolder: GridNodeHolder? = null
    private var actionSource: IActionSource? = null
    private var returnBuffer: Object2LongOpenHashMap<AEKey?> = Object2LongOpenHashMap<AEKey?>()

    @Persisted
    private var internalBuffer: KeyStorage? = null

    @Persisted
    private var internalTankBuffer: KeyStorage? = null

    init {
        this.nodeHolder = createNodeHolder()
        this.actionSource = IActionSource.ofMachine { nodeHolder!!.getMainNode().node }
    }

    protected fun createNodeHolder(): GridNodeHolder = GridNodeHolder(this)

    override fun getMainNode(): IManagedGridNode = nodeHolder!!.getMainNode()

    override fun isOnline(): Boolean = isMachineOnline

    override fun setOnline(online: Boolean) {
        isMachineOnline = online
    }

    override fun onMainNodeStateChanged(reason: IGridNodeListener.State) {
        super.onMainNodeStateChanged(reason)
        this.updateInventorySubscription()
    }

    override fun updateInventorySubscription() {
        if (shouldSubscribe()) {
            autoIOSubs = subscribeServerTick(autoIOSubs) { this.autoIO() }
        } else if (autoIOSubs != null) {
            autoIOSubs!!.unsubscribe()
            autoIOSubs = null
        }
    }

    /**/
    // /////////////////////////// */ // ***** Machine LifeCycle ****//
    /**/
    // /////////////////////////// */
    override fun createInventory(vararg args: Any?): NotifiableItemStackHandler {
        this.internalBuffer = KeyStorage()
        return InaccessibleInfiniteHandler(this, internalBuffer!!)
    }

    override fun createTank(initialCapacity: Int, slots: Int, vararg args: Any?): NotifiableFluidTank {
        this.internalTankBuffer = KeyStorage()
        return InaccessibleInfiniteTank(this, internalTankBuffer!!)
    }

    override fun onLoad() {
        super.onLoad()
        if (isRemote) return
    }

    override fun onMachineRemoved() {
        val grid = mainNode.getGrid()
        if (grid != null) {
            if (!internalBuffer!!.isEmpty) {
                for (entry in internalBuffer!!) {
                    grid.storageService.inventory.insert(
                        entry.key,
                        entry.longValue,
                        Actionable.MODULATE,
                        actionSource,
                    )
                }
            }
            if (!internalTankBuffer!!.isEmpty) {
                for (entry in internalTankBuffer!!) {
                    grid.storageService.inventory.insert(
                        entry.key,
                        entry.longValue,
                        Actionable.MODULATE,
                        actionSource,
                    )
                }
            }
        }
    }

    /**/
    // /////////////////////////// */ // ********** Sync ME *********//
    /**/
    // /////////////////////////// */
    protected fun shouldSubscribe(): Boolean = isWorkingEnabled && isOnline && (!(internalBuffer!!.storage.isEmpty() && internalTankBuffer!!.storage.isEmpty()))

    override fun autoIO() {
        if (!this.shouldSyncME()) return
        if (this.updateMEStatus()) {
            val grid = mainNode.getGrid()
            if (grid != null) {
                if (!internalBuffer!!.isEmpty) {
                    internalBuffer!!.insertInventory(grid.storageService.inventory, actionSource)
                }
                if (!internalTankBuffer!!.isEmpty) {
                    internalTankBuffer!!.insertInventory(grid.storageService.inventory, actionSource)
                }
            }
            this.updateInventorySubscription()
        }
    }

    override fun onRotated(oldFacing: Direction, newFacing: Direction) {
        super.onRotated(oldFacing, newFacing)
        mainNode.setExposedOnSides(EnumSet.of(newFacing))
    }

    override fun isWorkingEnabled(): Boolean = true

    override fun createUIWidget(): Widget {
        val group = WidgetGroup(Position(0, 0))
        // ME Network status
        group.addWidget(
            LabelWidget(
                0,
                0,
            ) { if (this.isMachineOnline) "gtceu.gui.me_network.online" else "gtceu.gui.me_network.offline" },
        )

        group.addWidget(AEListGridWidget.Item(5, 20, 3, this.internalBuffer))
        group.addWidget(AEListGridWidget.Fluid(5, 80, 3, this.internalTankBuffer))
        return group
    }

    override fun isRemote(): Boolean = holder.level()?.isClientSide ?: true
}
